devito/visitors.py

```diff
@@ -363,9 +414,55 @@ def visit_Iteration(self, o, subs={}, offsets={}):
             subs[o.dim] = Symbol(vname)
             return o._rebuild(*nodes, index=vname)
 
-    def visit_Expression(self, o, subs={}, offsets={}):
-        # Collect all offsets used with a dimension
-        # TODO: Straight update is not really safe,
-        # since we're mapping into a list...
-        offsets.update(o.index_offsets)
+    def visit_Expression(self, o, subs={}, offsets=defaultdict(set)):
+        """Collect all offsets used with a dimension"""
+        for dim, offs in o.index_offsets.items():
+            offsets[dim].update(offs)
         return o
+
+
+class MergeOuterIterations(Transformer):
```

Comment:
Using this visitor is legal as long as there are no particular dependencies across the merged expressions. We know this is always true for us... see below for more info about this